lib/Expression_gen.ml

```diff
@@ -1,262 +1,313 @@
+  pf ppf "(stan::math::value(%a) %s stan::math::value(%a))" pp_expr lhs op
+    pp_expr rhs
 
-and first es = (List.nth_exn es 0)
+and pp_unary ppf fm es = pf ppf fm pp_expr (List.hd_exn es)
+and pp_binary ppf fm es = pf ppf fm pp_expr (first es) pp_expr (second es)
+and first es = List.nth_exn es 0
+and second es = List.nth_exn es 1
 
-and second es = (List.nth_exn es 1)
-
-and gen_scalar_binary scalar_fmt generic_fmt es =
-  gen_binary (if (is_scalar (first es) && is_scalar(second es)) then scalar_fmt else generic_fmt) es
+and pp_scalar_binary ppf scalar_fmt generic_fmt es =
+  pp_binary ppf
+    ( if is_scalar (first es) && is_scalar (second es) then scalar_fmt
+    else generic_fmt )
+    es
 
 (* assumes everything well formed from parser checks *)
-and gen_fun_app f es = match f with
-  | "And" -> gen_short_circuit_logical_op "&&" es
-  | "Or" -> gen_short_circuit_logical_op "||" es
-  | "PMinus" -> gen_unary (if is_scalar (List.hd_exn es) then "-%s" else "minus(%s)") es
-  | "PPlus" -> gen_unary "%s" es
-  | "Transpose" -> gen_unary (if (is_scalar (List.hd_exn es)) then "transpose(%s)"  else "%s") es
-  | "PNot" -> gen_unary "logial_negation(%s)" es
-  | "Minus" -> gen_scalar_binary "(%s - %s)" "subtract(%s, %s)" es
-  | "Plus" -> gen_scalar_binary "(%s + %s)" "add(%s, %s)" es
-  | "Times" -> gen_scalar_binary "(%s * %s)" "multiply(%s, %s)" es
-  | "Divide" -> if (is_matrix (second es) && (is_matrix (first es) || is_row_vector (first es)))
-    then gen_binary "mdivide_right(%s, %s)" es
-    else gen_scalar_binary "(%s / %s)" "divide(%s, %s)" es
-  | "Modulo" -> gen_binary "modulus(%s, %s)" es
-  | "LDivide" -> gen_binary "mdivide_left(%s, %s)" es
-  | "EltTimes" -> gen_scalar_binary "(%s * %s)" "elt_multiply(%s, %s)" es
-  | "EltDivide" -> gen_scalar_binary "(%s / %s)" "elt_divide(%s, %s)" es
-  | "Pow" -> gen_binary "pow(%s, %s)" es
-  | "Equals" -> gen_binary "logical_eq(%s, %s)" es
-  | "NEquals" -> gen_binary "logical_neq(%s, %s)" es
-  | "Less" -> gen_binary "logical_lt(%s, %s)" es
-  | "Leq" -> gen_binary "logical_lte(%s, %s)" es
-  | "Greater" -> gen_binary "logical_gt(%s, %s)" es
-  | "Geq" -> gen_binary "logical_gte(%s, %s)" es
-  | "lmultiply" -> gen_binary "multiply_log(%s, %s)" es
-  | "lchoose" -> gen_binary "binomial_coefficient_log(%s, %s)" es 
-  | "target" -> "get_lp(lp__, lp_accum__)"
-  | "get_lp" -> "get_lp(lp__, lp_accum__)"
-  | "max" -> if (List.length es = 2) then gen_binary "std::max(%s, %s)" es
-    else gen_ordinary_function f es
-  | "min" -> if (List.length es = 2) then gen_binary "std::min(%s, %s)" es
-    else gen_ordinary_function f es 
-  | "ceil" -> if (is_scalar (first es)) then gen_unary "std::ceil(%s)" es
-    else gen_ordinary_function f es
-  | _ -> gen_ordinary_function (stan_namespace_qualify f) es 
-
-and gen_ordinary_function f es =
-  sprintf "%s(%s)" f (gen_sep (gen_exprs es) (gen_extra_fun_args f))
-
-and gen_expr (e : expr_typed_located) =
+and gen_fun_app ppf ut f es =
```

Comment:
I'll give it a shot. This may just be a terrible pattern we should change somehow. @enetsee do you have any advice for how to do these large switches on function names?

Reply:
There might be a few things that could help. 

First, I would consider splitting out the three maps you are creating so we have a separate function for each of:

- `operator_functions`
- `misc_special_math_fns`
- `mir_special_functions`

For the `operator_functions`, since there is a fixed set represented at an ADT, I would also suggest creating a function `operator_from_string` with type `string -> operator option`.

For the other functions, since you have enumerated the possible string representation, I would probably have them as union types and create `from_string` functions as above.

I would also pull the default case out.

We end up with four functions, three of which return options and the last (the default) which always gives us a function to pretty print with. We can them combine them in the top level `gen_oerator_app` function.

```
and operator_from_string str = 
  match str with
  | "+" -> Some Ast.Plus 
  | ...  ->  etc

and gen_operator_app  f  = 
  operator_from_string f
  |> Option.map
      ~f:(fun op -> 
        match op with 
        | Ast.PMinus -> 
              (fun ppf es -> 
                pp_unary ppf
                  (if is_scalar (List.hd_exn es) then "-%a" else "minus(%a)")
                  es
              )
          
        | PPlus -> 
            (fun ppf es -> pp_unary ppf "%a" es)

        | ... -> etc
      )
 
and gen_special_math_fn_app f = 
  ... 

and gen_mir_special_fn_app ut f = 
  ... 

(* There might already be a function in `Option` with this sig *)
and or_else ~default option = 
  match option with
  | Some _ -> option 
  | _ -> default

and default_fn_app f ppf es = 
  pp_ordinary_fn ppf (stan_namespace_qualify f) es

(* assumes everything well formed from parser checks *)
and gen_fun_app ppf ut f es =
  let f = 
    gen_operator_app f
    |> or_else ~default:(gen_special_math_fn_app f)
    |> or_else ~default:(gen_mir_special_fn_app ut f)
    |> Option.value ~default:(default_fn_app f )
  in 
    f ppf es
```

My personal preference is not to have lots of big, nested functions but rather to hide them with the `.mli` files and then compose them and expose the top level function.

Reply:
Ah, I was trying to figure out WTF `first_some` did from the doc string the other day. This use case makes sense. Thank you!!! I will refactor it like this.

Reply:
Thans, @enetsee.  Me, too!

> My personal preference is not to have lots of big, nested functions but rather to hide them with the .mli files and then compose them and expose the top level function.

Reply:
> For the operator_functions, since there is a fixed set represented by an ADT, I would also suggest creating a function operator_from_string with type string -> operator option.

Is this just in case we need it in the future?

Reply:
I was more thinking along the lines of creating functions that do one thing  and then composing, but, yes, that usually helps if/when other things come up in the future

Reply:
I guess I didn't see where we needed it in composition at the moment?

Reply:
OH sorry, I see it now, I'm dumb. Thanks!